birdset/modules/finetune_module.py

```diff
@@ -0,0 +1,95 @@
+from birdset.modules.base_module import BaseModule
+from functools import partial
+from typing import Callable, Literal, Type, Optional, Union
+from torch.nn import CrossEntropyLoss
+from torch.nn.modules.loss import _Loss
+from torch.optim import AdamW, Optimizer
+from birdset.configs import NetworkConfig, LoggingParamsConfig, LRSchedulerConfig, MulticlassMetricsConfig, MultilabelMetricsConfig, MultilabelMetricsConfig as MetricsConfig
+from birdset.datamodule.embedding_datamodule import EmbeddingModuleConfig
+from birdset.utils import pylogger
+
+import torch
+import math
+
+log = pylogger.get_pylogger(__name__)
+class FinetuneModule(BaseModule):
+    def __init__(
+            self,
+            network: NetworkConfig = NetworkConfig(),
+            output_activation: Callable[[torch.Tensor], torch.Tensor] = partial(
+                torch.softmax,
+                dim=1
+            ),
+            loss: _Loss = CrossEntropyLoss(),
+            optimizer: partial[Type[Optimizer]] = partial(
+                AdamW,
+                lr=1e-5,
+                weight_decay=0.01,
+            ),
+            lr_scheduler: Optional[LRSchedulerConfig] = LRSchedulerConfig(),
+            metrics: MulticlassMetricsConfig | MultilabelMetricsConfig = MulticlassMetricsConfig(),
+            logging_params: LoggingParamsConfig = LoggingParamsConfig(),
+            num_epochs: int = 50,
+            len_trainset: int = 13878, # set to property from datamodule
+            batch_size: int = 32,
+            task: Literal['multiclass', 'multilabel'] = "multiclass",
+            num_gpus: int = 1,
+            pretrain_info = None,
+            embedding_model: EmbeddingModuleConfig = EmbeddingModuleConfig(), # Model for extracting the embeddings
+            ):
+        super().__init__(
+            network = network,
+            output_activation=output_activation,
+            loss=loss,
+            optimizer=optimizer,
+            lr_scheduler=lr_scheduler,
+            metrics=metrics,
+            logging_params=logging_params,
+            num_epochs=num_epochs,
+            len_trainset=len_trainset,
+            batch_size=batch_size,
+            task=task,
+            num_gpus=num_gpus,
+            pretrain_info = pretrain_info,
+        )
+        self.embedding_model = embedding_model.model
+        self.sampling_rate = embedding_model.sampling_rate
+        self.max_length = embedding_model.length  
+        log.info(f"FINETUNING using embedding model:{embedding_model.model_name} (Sampling Rate:{self.sampling_rate}, Window Size:{self.max_length})")
+
+    def configure_optimizers(self):
+        self.optimizer = self.optimizer(list(self.model.parameters())+list(self.embedding_model.parameters())) #! Changed this
+        if self.lr_scheduler is not None:
+            # TODO: Handle the case when we do not want warmup
```

Comment:
If you do not attend to solve those TODOs in this PR, please create an issue and add a link to the issue in the comments.